ethcore/sync/src/chain/mod.rs

```diff
@@ -855,14 +853,29 @@ impl ChainSync {
 	fn collect_blocks(&mut self, io: &mut SyncIo, block_set: BlockSet) {
 		match block_set {
 			BlockSet::NewBlocks => {
-				if self.new_blocks.collect_blocks(io, self.state == SyncState::NewBlocks) == Err(DownloaderImportError::Invalid) {
+				if self.new_blocks.collect_blocks(io, self.state == SyncState::NewBlocks) == DownloadAction::Reset {
+					trace!(target: "sync", "Restarting NewBlocks download");
 					self.restart(io);
```

Comment:
In the case of a successfully completed sync round, this probably shouldn't `restart()`? Why not handle it similar to OldBlocks by expiring peer requests and calling `reset()`?

Reply:
Indeed we shouldn't. I've reverted the change which returns `DownloadAction::Reset` in case of successful round completion. 

I also agree this could be unified. My thinking for this PR was to minimize changes that might break `NewBlocks` sync, and just get something to fix ancient blocks.